### PR TITLE
Fix issue67

### PR DESCRIPTION
**Pull request by mmatera**
From #67, some issues related both to ``FindRoot`` and ``NIntegrate`` become apparent:  
* `D[NIntegrate[...]]` - like were not evaluated.
* ``FindRoot`` didn't handle functions whose derivatives can not be explicitly evaluated.
* ``$ScriptCommandLine`` now has the name of the script as a first element (like in WMA).

### CODE REVIEW

File: mathics/builtin/numbers/calculus.py [hunk 1]
```diff
@@ -1287,24 +1320,104 @@ def find_root_secant(f, x0, x, opts, evaluation) -> (Number, bool):
...
+        Check if val is zero upto the precision and accuracy goals
+        """
+        if not val.is_numeric():
+            return False
+        if val.is_zero:
+            return True
+        if acc_goal:
+            if prec_goal:
+                eps = apply_N(
+                    Expression(
```
Comment (TiagoCavalcante) [inline mathics/builtin/numbers/calculus.py:1355]: Maybe create a `SymbolLog`?

**Review by TiagoCavalcante — APPROVED** [review]
LGTM

File: mathics/builtin/system.py [hunk 2]
```diff
@@ -345,10 +345,9 @@ def evaluate(self, evaluation):
         except ValueError:
             # not run in script mode
             return Expression(SymbolList)
-
-        return Expression(
-            SymbolList, *(String(arg) for arg in sys.argv[dash_index + 1 :])
-        )
+        scriptname = "" if dash_index == 0 else sys.argv[dash_index - 1]
```
Comment (rocky) [inline mathics/builtin/system.py:348]: This seems totally independent. What's up here? 

This probably should go into a separate PR.

File: CHANGES.rst [hunk 3]
```diff
@@ -4,7 +4,11 @@ CHANGES
 New Builtins
 ============
 * ``TraceEvaluation[]`` shows expression name calls and return values of it argument. The variable ``$TraceEvalution`` when set True will show all expression evaluations.
-
+* Now, ``D`` can act over ``NIntegrate``.
+* ``FindRoot`` was improved for supporting numerical derivatives.
+* ``FindRoot`` (for the ``newton`` method) partially supports ``EvaluationMonitor`` and ``StepMonitor`` options. 
+* ``$ScriptCommandLine`` now has the filename of the script as its first element.
```
Comment (rocky) [inline CHANGES.rst:10]: See below for `$ScriptCommandLine` change.

File: mathics/builtin/assignments/internals.py [hunk 4]
```diff
@@ -153,7 +154,7 @@ def find_tag_and_check(lhs, tags, evaluation):
 
 
 def unroll_patterns(lhs, rhs, evaluation):
-    if type(lhs) is Symbol:
+    if isinstance(lhs, Atom):
```
Comment (rocky) [inline mathics/builtin/assignments/internals.py:157]: Is the idea that unrolling works if we have number as well and we weren't catching this?

File: mathics/builtin/numbers/calculus.py [hunk 5]
```diff
@@ -58,6 +66,33 @@
...
+def get_accuracy_and_prec(opts: dict, evaluation: "Evaluation"):
+    """
+    Looks at an opts dictionary and tries to determine the numeric values of
+    Accuracy and Precision goals. If not available, returns None.
+    """
+    acc_goal = opts.get("System`AccuracyGoal", None)
+    if acc_goal:
+        acc_goal = apply_N(acc_goal, evaluation)
+        if acc_goal is SymbolAutomatic:
+            acc_goal = from_python(9.0)
```
Comment (rocky) [inline mathics/builtin/numbers/calculus.py:78]: Where does the number 9.0 come from?

File: mathics/builtin/numbers/calculus.py [hunk 6]
```diff
@@ -1407,10 +1535,30 @@ class FindRoot(Builtin):
...
     methods = {
+        "Automatic": find_root_newton,
         "Newton": find_root_newton,
         "Secant": find_root_secant,
     }
 
+    @staticmethod
+    def determine_epsilon(x0, options, evaluation):
+        """Determine epsilon  from a reference value, and from the accuracy and the precision goals"""
+        acc_goal, prec_goal = get_accuracy_and_prec(options, evaluation)
```
Comment (rocky) [inline mathics/builtin/numbers/calculus.py:1546]: My own personal opinion this is the kind of unnecessary coding is that I see a lot of which I find penny wise and pound foolish.

There are two independent parameters, acc_goal and prec_goal and each contributes to the returned value. We have this two if/else's embedded into in another if/else each with some sort of repeated expression given in Mathics expression - like format (i.e. is a little harder to understand than say a Python expression). 

A straightforward and easy-to-understand way to code this is simply not nest and duplicate expressions 

For example: 

```
# from_python(1e-10) is computed once outside and given a name and comment that describes its meaning 

if not (pres_goal or acc_goal): 
     return WHATEVER_FROM_1E_MINUS10_MEANS

eps = abs(x0) *  Integer10 ** (-prec_goal) if pres_goal else Integer0
if acc_goal:
   eps += Integer10 ** (-acc_goal)
return = apply_N(eps)
```

File: mathics/builtin/numbers/calculus.py [hunk 7]
```diff
@@ -1315,12 +1430,23 @@ def sub(evaluation):
...
             evaluation.message("FindRoot", "nnum", x, x0)
             return x0, False
-        # TODO: use Precision goal...
-        if x1 == x0:
-            break
-        x0 = apply_N(x1, evaluation)
-        # N required due to bug in sympy arithmetic
-        count += 1
+
+        # Check convergency:
```
Comment (rocky) [inline mathics/builtin/numbers/calculus.py:1434]: convergency -> convergence

File: CHANGES.rst [hunk 8]
```diff
@@ -4,7 +4,11 @@ CHANGES
 New Builtins
 ============
 * ``TraceEvaluation[]`` shows expression name calls and return values of it argument. The variable ``$TraceEvalution`` when set True will show all expression evaluations.
-
+* Now, ``D`` can act over ``NIntegrate``.
```
Comment (rocky) [inline CHANGES.rst:7]: Let's please separate the ``D`` change in the PR from the ``FindRoot`` change.

**Comment by rocky** [comment]
A thought I keep having after looking over stuff like this. Although I think we should merge this, I keep thinking that what we should be focusing on instead is how to hook into Sympy to do Newton's (or someone else's method) for finding Roots and Minima.

File: mathics/builtin/system.py (hunk 2 above)
Comment (mmatera) [reply mathics/builtin/system.py:348]: @rocky,
the issue contained several aspects. 
- How `FindRoot` handles expressions which can not have an analytic derivative
- How to evaluate derivatives of `NIntegral`
- How we are processing command line parameters.

So I put here all together to see if it solves the test case proposed.

In this particular module, I proposed a fixed for the last one.

**Comment by mmatera** [comment]
> A thought I keep having after looking over stuff like this. Although I think we should merge this, I keep thinking that what we should be focusing on instead is how to hook into Sympy to do Newton's (or someone else's method) for finding Roots and Minima.

Rocky, to use libraries to implement these calculus is  work in progress. Now I am implementing an interface for Scipy, but I wanted to have some feedback from this PR to polish and publish it.

File: mathics/builtin/numbers/calculus.py (hunk 6 above)
Comment (mmatera) [reply mathics/builtin/numbers/calculus.py:1546]: @rocky,  OK, I like it! And no, it was not about performance, it was just about having a simplest routine that came to my mind that provides the output I needed. This one is much better!

File: mathics/builtin/numbers/calculus.py (hunk 5 above)
Comment (mmatera) [reply mathics/builtin/numbers/calculus.py:78]: In WMA, `Chop` by default round numbers to 10 significant digits. Here I just look for a number around 10 that provides results accurate enough to pass the tests. In any case, for a final version, should let to the different methods that define the defaults.

File: mathics/builtin/assignments/internals.py (hunk 4 above)
Comment (mmatera) [reply mathics/builtin/assignments/internals.py:157]: This is something that found by chance: suppose that by mistake, you evaluate
`7/; F[a]:=7`
in master, it raises an unhandled exception. With this change, just an error message.

File: mathics/builtin/system.py (hunk 2 above)
Comment (rocky) [reply mathics/builtin/system.py:348]: Ok. But let's me clear. Users and issues will sometimes be something like: please fix everything that I just encountered.

We should try to split up the issues so that each covers one thing so that thing can be tracked. And similarly in PRs it would be nice to split separate problems into separate PRs for the same reason. They will get through faster. 

And in 6 months when we are looking over looking at a particular line of code and are wondering why it is there and then look at the commit or merge associated with it, it reflects one idea, and not several.

**Comment by mmatera** [comment]
split and enlarged in  #129, #131, #133
